File: cli-implementations/js/crystallizationManager.ts
import fs from 'fs';
import path from 'path';
import yargs from 'yargs';
import { hideBin } from 'yargs/helpers';
import {
  loadData,
  saveData,
  updateKPI,
  addTask,
  updateAttributes,
  Data,
} from '../../src/core';

yargs(hideBin(process.argv))
  .command(
    'add-task',
    'Add new task',
    (y) =>
      y
        .option('id', { type: 'string', demandOption: true })
        .option('title', { type: 'string', demandOption: true })
        .option('complexity', { type: 'string' })
        .option('tags', { type: 'string' }),
    (argv) => {
      const data = loadData();
      const ok = addTask(data, {
        id: String(argv.id),
        title: String(argv.title),
        complexity: argv.complexity ? String(argv.complexity) : undefined,
        tags: argv.tags
          ? String(argv.tags)
              .split(',')
              .map((t) => t.trim())
          : undefined,
      });
      if (!ok) {
        console.error('Task already exists');
        return;
      }
      saveData(data);
      console.log('Task added');
    }
  )
  .command(
    'update-kpi',
    'Update KPI values for task',
    (y) =>
      y
        .option('id', { type: 'string', demandOption: true })
        .option('metrics', { type: 'string', demandOption: true }),
    (argv) => {
      const data = loadData();
      const task = data.tasks.find((t) => t.id === argv.id);
      if (!task) {
        console.error('Task not found');
        return;
      }
      let metrics: Record<string, number> = {};
      try {
        metrics = JSON.parse(String(argv.metrics));
      } catch {
        console.error('Invalid metrics JSON');
        return;
      }
      updateKPI(task, metrics, data.kpi_definitions);
      saveData(data);
      console.log(
        task.status === 'diamond'
          ? '💎 Diamond achieved!'
          : `Progressed to iteration ${task.iteration}.`
      );
    }
  )
  .command(
    'update-attrs',
    'Update task complexity and tags',
    (y) =>
      y
        .option('id', { type: 'string', demandOption: true })
        .option('complexity', { type: 'string' })
        .option('tags', { type: 'string' }),
    (argv) => {
      const data = loadData();
      const task = data.tasks.find((t) => t.id === argv.id);
      if (!task) {
        console.error('Task not found');
        return;
      }
      updateAttributes(task, {
        complexity: argv.complexity ? String(argv.complexity) : undefined,
        tags: argv.tags
          ? String(argv.tags)
              .split(',')
              .map((t) => t.trim())
          : undefined,
      });
      saveData(data);
      console.log('Task attributes updated');
    }
  )
  .command(
    'level',
    'Show crystallization level for task',
    (y) => y.option('id', { type: 'string', demandOption: true }),
    (argv) => {
      const data = loadData();
      const task = data.tasks.find((t) => t.id === argv.id);
      if (!task) {
        console.error('Task not found');
        return;
      }
      console.log(task.final_score ?? '0');
    }
  )
  .command(
    'update-core',
    'Update core principles',
    (y) => y.option('principles', { type: 'string', demandOption: true }),
    (argv) => {
      const data = loadData();
      data.core_principles = String(argv.principles)
        .split(',')
        .map((p) => p.trim());
      data.core_version += 1;
      saveData(data);
      console.log('Core principles updated');
    }
  )
  .command(
    'average',
    'Show average crystallization level',
    () => {},
    () => {
      const data = loadData();
      const scores = data.tasks.map((t) => t.final_score ?? 0);
      const avg = scores.reduce((a, b) => a + b, 0) / (scores.length || 1);
      console.log(avg.toFixed(2));
    }
  )
  .command(
    'list-diamonds',
    'List all tasks in diamond state',
    () => {},
    () => {
      const data = loadData();
      const diamonds = data.tasks.filter((t) => t.status === 'diamond');
      if (diamonds.length === 0) {
        console.log('No diamond tasks');
        return;
      }
      diamonds.forEach((t) => {
        console.log(`${t.id}: ${t.title}`);
      });
    }
  )
  .command(
    'list-funcs',
    'Scan project functions and store the most common',
    () => {},
    async () => {
      const root = path.resolve(__dirname, '..', '..');
      const tsFiles: string[] = [];
      const exclude = ['node_modules', '.git'];

      function gather(dir: string) {
        for (const file of fs.readdirSync(dir)) {
          if (exclude.includes(file)) continue;
          const full = path.join(dir, file);
          const stat = fs.statSync(full);
          if (stat.isDirectory()) {
            gather(full);
          } else if (full.endsWith('.ts')) {
            tsFiles.push(full);
          }
        }
      }

      gather(root);

      const cachePath = path.resolve(root, '.listfuncs_cache.json');
      let cache: Record<
        string,
        { mtime: number; counts: Record<string, number> }
      > = {};
      if (fs.existsSync(cachePath)) {
        try {
          cache = JSON.parse(fs.readFileSync(cachePath, 'utf-8'));
        } catch {
          cache = {};
        }
      }

      const counts: Record<string, number> = {};
      const newCache: typeof cache = {};
      const reg = /function\s+(\w+)|const\s+(\w+)\s*=\s*\(/g;

      await Promise.all(
        tsFiles.map(async (f) => {
          const stat = fs.statSync(f);
          const cached = cache[f];
          if (cached && cached.mtime === stat.mtimeMs) {
            Object.entries(cached.counts).forEach(([name, c]) => {
              counts[name] = (counts[name] || 0) + c;
            });
            newCache[f] = cached;
            return;
          }

          const content = await fs.promises.readFile(f, 'utf-8');
          const fileCounts: Record<string, number> = {};
          let m: RegExpExecArray | null;
          while ((m = reg.exec(content))) {
            const name = m[1] || m[2];
            if (!name) continue;
            fileCounts[name] = (fileCounts[name] || 0) + 1;
            counts[name] = (counts[name] || 0) + 1;
          }
          newCache[f] = { mtime: stat.mtimeMs, counts: fileCounts };
        })
      );

      fs.writeFileSync(cachePath, JSON.stringify(newCache, null, 2));

      const entries = Object.entries(counts).sort((a, b) => b[1] - a[1]);

      if (entries.length) {
        const result = entries[0];
        const resultPath = path.resolve(root, 'most_common_function.txt');
        fs.writeFileSync(resultPath, `${result[0]}\n`);
        console.log(`Most common function: ${result[0]} (${result[1]})`);
        console.log(`Result saved to ${resultPath}`);
      } else {
        console.log('No functions found');
      }
    }
  )
  .command(
    'list-tasks',
    'List all tasks with statuses',
    () => {},
    () => {
      const data = loadData();
      if (data.tasks.length === 0) {
        console.log('No tasks found');
        return;
      }
      data.tasks.forEach((t) => {
        console.log(`${t.id}: ${t.title} [${t.status ?? 'unknown'}]`);
      });
    }
  )
  .command(
    'stats',
    'Show summary statistics',
    () => {},
    () => {
      const data = loadData();
      const statusCounts: Record<string, number> = {};
      data.tasks.forEach((t) => {
        const status = t.status ?? 'unknown';
        statusCounts[status] = (statusCounts[status] || 0) + 1;
      });
      const avg =
        data.tasks.reduce((sum, t) => sum + (t.final_score ?? 0), 0) /
        (data.tasks.length || 1);
      console.log('Task status counts:');
      Object.entries(statusCounts).forEach(([k, v]) =>
        console.log(`  ${k}: ${v}`)
      );
      console.log(`Average score: ${avg.toFixed(2)}`);
    }
  )
  .command(
    'init',
    'Initialize crystallization in current repo',
    () => {},
    () => {
      const target = path.resolve(process.cwd(), 'crystallization.json');
      if (fs.existsSync(target)) {
        console.error('crystallization.json already exists');
        return;
      }
      const template: Data = {
        core_version: 1,
        core_principles: ['Iterative improvement', 'KPI-driven development'],
        tasks: [],
        kpi_definitions: [
          { key: 'cycle_time_days', title: 'Cycle Time (days)', threshold: 3 },
          { key: 'code_coverage', title: 'Code Coverage', threshold: 0.8 },
        ],
      };
      fs.writeFileSync(target, JSON.stringify(template, null, 2));
      console.log('Initialized crystallization.json');
    }
  )
  .demandCommand()
  .help()
  .strict()
  .parse();

File: cli-implementations/python/crystallization_manager.py
import json
import argparse
from pathlib import Path

DATA_FILE = Path(__file__).resolve().parents[2] / 'crystallization.json'


def load_data():
    with open(DATA_FILE, 'r', encoding='utf-8') as f:
        return json.load(f)


def save_data(data):
    with open(DATA_FILE, 'w', encoding='utf-8') as f:
        json.dump(data, f, ensure_ascii=False, indent=2)


def add_task(args):
    data = load_data()
    if any(t['id'] == args.id for t in data['tasks']):
        print('Task already exists')
        return
    data['tasks'].append({
        'id': args.id,
        'title': args.title,
        'status': 'backlog',
        'iteration': 0,
        'kpi_history': []
    })
    save_data(data)
    print('Task added')


def update_kpi(args):
    data = load_data()
    task = next((t for t in data['tasks'] if t['id'] == args.id), None)
    if not task:
        print('Task not found')
        return
    iter_num = task['iteration'] + 1
    task['iteration'] = iter_num
    task.setdefault('kpi_history', []).append({
        'iteration': iter_num,
        'score': args.score,
        'notes': args.notes
    })
    task['final_score'] = args.score
    save_data(data)
    print('KPI updated')


def level(args):
    data = load_data()
    task = next((t for t in data['tasks'] if t['id'] == args.id), None)
    print(task.get('final_score', 0) if task else 0)


def update_core(args):
    data = load_data()
    data['core_principles'] = [p.strip() for p in args.principles.split(',')]
    data['core_version'] += 1
    save_data(data)
    print('Core principles updated')


def average(args):
    data = load_data()
    scores = [t.get('final_score', 0) for t in data['tasks']]
    avg = sum(scores) / (len(scores) or 1)
    print(f"{avg:.2f}")


def init_repo(args):
    if DATA_FILE.exists():
        print('crystallization.json already exists')
        return
    template = {
        'core_version': 1,
        'core_principles': ['Iterative improvement', 'KPI-driven development'],
        'kpi_definitions': [
            {'key': 'cycle_time_days', 'title': 'Cycle Time (days)', 'threshold': 3},
            {'key': 'code_coverage', 'title': 'Code Coverage', 'threshold': 0.8}
        ],
        'tasks': []
    }
    with open(DATA_FILE, 'w', encoding='utf-8') as f:
        json.dump(template, f, ensure_ascii=False, indent=2)
    print('Initialized crystallization.json')


parser = argparse.ArgumentParser(description='Crystallization CLI (Python)')
sub = parser.add_subparsers(dest='command')

add = sub.add_parser('add-task')
add.add_argument('--id', required=True)
add.add_argument('--title', required=True)
add.set_defaults(func=add_task)

kpi = sub.add_parser('update-kpi')
kpi.add_argument('--id', required=True)
kpi.add_argument('--score', type=float, required=True)
kpi.add_argument('--notes', default='')
kpi.set_defaults(func=update_kpi)

lvl = sub.add_parser('level')
lvl.add_argument('--id', required=True)
lvl.set_defaults(func=level)

core = sub.add_parser('update-core')
core.add_argument('--principles', required=True)
core.set_defaults(func=update_core)

avg = sub.add_parser('average')
avg.set_defaults(func=average)

init_p = sub.add_parser('init')
init_p.set_defaults(func=init_repo)

args = parser.parse_args()
if hasattr(args, 'func'):
    args.func(args)
else:
    parser.print_help()
